### DISCUSSION TITLE
How to use 'ada::result' efficiently.

### DISCUSSION THREAD

**Discussion by lemire** [ada-url/ada#200] [General]
Our parse function produces a 'ada::result' which is an alias for an `std::expected<ada::url,...>` (from C++23).

Our main objective is to make it very hard for users to access the result without first checking for an error condition. It is a common problem that people make assumptions (the URL is always valid) only to find the assumption violated in production. We want to help users do the right thing naturally.

You can use it as like so...

```C++
ada::result url = ada::parse(...);
if(url) {
  url->get_scheme();// safe and won't throw
}
```

If you forget to check...

```C++
ada::result url = ada::parse(.bad url.);
url->get_scheme();// throws!!!
```

It is unsafe code.

If you need to access directly the `ada::url` instance, you can do this...

```C++
ada::result url = ada::parse(...);
if(url) {
  ada::url u = url.value(); // get a copy
  // OR: ada::url u = *url; // get a copy
}
```

However, creating a copy is wasteful. 

You may grab a reference which is efficient:

```C++
ada::result url = ada::parse(...);
if(url) {
  ada::url& u = url.value(); // no copy is made
}
```

*Returning from functions*:

If you have a function returning an ada::url, then you won't copy...

```C++

ada::url get_url(std::string_view url) {
  return ada::parse(...).value(); // no copy made
}
```

However, it will throw on error.

You can do it this way...

```C++

ada::url get_url(std::string_view url) {
  return ada::parse(...).value_or(ada::url());
}
```

But you will be creating a *second* struct (default ada::url) so it is not free.

You might as well pass around `ada::result` instances unless you can tell that, for sure, that the url is valid.

```C++

ada::result get_url(std::string_view url) {
  return ada::parse(...); //  efficient
}
```

*Performance tip*: Avoid copying by moving the result. After checking the value is valid, you can move it to a variable. It is efficient.

```C++
ada::result url = ada::parse(...);
if(url) {
  ada::url u = std::move(*url); // move to u
  // OR: ada::url u = std::move(url.value()); // move to u
}
```

*Pointers*: 

If you just need a pointer to the ada::url value contained in the ada::result, you can do this:

```C++
ada::result url = ada::parse(...);

if(url) {
  ada::url* u = &*url;
  // OR: ada::url* u = &url.value();
}
```

It should not entail any copying and should be very efficient. Note that the pointer is only valid as long as `ada::result url`is alive.

A pointer is likely even faster than a move!!!

**Comment by anonrig** [comment]
cc @ronag @addaleax @jasnell what do you think?

**Reply by ronag** [reply]
What is the question? LGTM
